Update table_to_annotations.py

Bankso opened the pull request:
@vpchung when I went to run this, I realized that I did not test thoroughly enough before opening my PR.

There were two things I did that were causing some weirdness:
- I was applying annotations based on the number of entries in the metadata tables (when it should be the number of files in the dataset)
- I wrote everything using (record key, file id) tuples/dictionaries, when I should have been using (file id, record key). The record keys aren't necessarily unique, so I was overwriting the values. I didn't notice it until I went to apply individual metadata and the script was stopping at 84 entries (the # of rows in the individual table), when it should have been 253.

I added changes that appear to have fixed things. As always, I'm not sure if this is the "correct" way of doing it, but it does seem to be working as expected.

Bankso committed:
Update table_to_annotations.py

Issue: annotations were being applied based on the number of table rows and dictionary keys were being overwritten

Solution: Invert file_tuple to be (file id, biospecimen key); add a for loop to collect annotations functions, which will create and apply annotations for each file in file_tuples dict; add logic to apply annotations if the primary key of the table row matches the foreign key extracted from current annotations

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -73,15 +73,15 @@ def collect_fileview_annotations(syn, files: list, fileview_id: str) -> dict:
         id_key_tuples.append(id_key_tuple)
     
-    file_tuples = dict([tup for tup in id_key_tuples if tup[1] in files])
+    file_tuples = dict([tup for tup in id_key_tuples if tup[0] in files])
```
Whoops!  Sorry I missed this in the previous PR.  This is definitely on the right track, but I think you may have over-complicated it _juuuust_ a little bit 😉  How about using a dictionary comprehension instead?

```python
fileview_table = get_table(syn, fileview_id, fileview_columns)
file_biospecimen_mapping = {
    row["FileView_id"]: row["Biospecimen Key"]
    for _, row in fileview_table.iterrows()
    if row["FileView_id"] in files
}
```

Since `fileview_table` is a pandas dataframe, leveraging its methods is also an option!  For example, here I would first filter for rows where `FileView_id` is found in `files` by using `isin()`, then I would set `FileView_id` as the index (since that's what we want as our key).  Then I would use `to_dict()` to create a dictionary for each column in the dataframe as the value.  Finally, I would select which of the dictionaries I'm interested in, in this case, we want the dictionary where it's FileView_id -> Biospecimen Key:

```python
fileview_table = get_table(syn, fileview_id, fileview_columns)
file_biospecimen_mapping = (
    fileview_table[fileview_table["FileView_id"].isin(files)]
    .set_index("FileView_id")  # key: file synID
    .to_dict()["Biospecimen Key"]  # value: biospecimen
)
```

vpchung replied:
^^^ Ah, quick note here: I'm using `file_biospecimen_mapping` as a variable name over `file_tuples`, as the latter name is a bit misleading.

vpchung replied:
Also: it won't let me highlight L67 above above, but if I understand your proposed change correctly, the function will now:

`return a File Synapse Id: Biospecimen Key dictionary`

Bankso replied:
Oh wow, I did overcomplicate it! This makes sense, thank you! I like the dictionary comprehension approach!

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -73,15 +73,15 @@ def collect_fileview_annotations(syn, files: list, fileview_id: str) -> dict:
             
 
-def collect_biospecimen_annotations(syn, file_tuples: dict, specimen_info_tuple: tuple[str, str, str]) -> tuple[dict, dict]:
+def collect_biospecimen_annotations(syn, file_tuples: dict, specimen_info_tuple: tuple[str, str, str]) -> tuple[dict, dict, dict]:
```
nit: rename `file_tuples` so that it accurately reflects what kind of data it's holding, which as the type-hint suggests, a dictionary.

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -95,25 +95,26 @@ def collect_biospecimen_annotations(syn, file_tuples: dict, specimen_info_tuple:
     
     print(f"Biospecimen annotations applied to {count} entities")
-    return dict(individual_tup_list), dict(model_tup_list)
+    return file_tuples, dict(individual_tup_list), dict(model_tup_list)
```
I can't highlight L92-93, but one quick suggestion here is to use dictionaries from the get-go rather than using a list of tuples and converting.  Something like this:

```python
individual_map = {}
...
for ...:
    individual_map[file] = individual_id
...
return (file_tuples, individual_map, ...)
```

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -95,25 +95,26 @@ def collect_biospecimen_annotations(syn, file_tuples: dict, specimen_info_tuple:
+                    model_tup_list.append(model_tup)
+                    biospecimen_annotations = list(zip(column_list, list(row)))
+                    apply_annotations_to_entity(syn, component, file, biospecimen_annotations)
+                    count += 1
```
I would take advantage of the pandas methods here for efficiency (since nested `for` loops can get pretty slow once the iterables are bigger):

```python
# Get the metadata table then set the index as Biospecimen_id right away. 
# Indexing will make the lookups much faster!
data_table = get_table(syn, table_id, column_list).set_index("Biospecimen_id")

# You mentioned that the record keys might not be unique, so first converting values()
# into a set would remove those dups.
biospecimen_ids = set(file_tuples.values())

# Filter the data table so that you only get metadata for the relevant biospecimen IDs.
filtered_metadata = data_table[data_table.index.isin(biospecimen_ids)]

# Since you're interested in both the keys and values of the dictionary, I suggest
# using items() here so that you'll get both into named variables right away.
for file_id, biospecimen_key in file_tuples.items():
    metadata = filtered_metadata.loc[biospecimen_key]
    individual_id = metadata["Individual Key"]
    model_id = metadata["Model Key"]
    biospecimen_annotations = list(zip(column_list, metadata.tolist()))
    ...
```

vpchung replied:
^^^ thinking on this more, maybe we should add an additional check for robustness:

```python
for file_id, biospecimen_key in file_tuples.items():
    if biospecimen_key in filtered_metadata:
        metadata = filtered_metadata.loc[biospecimen_key]
        individual_id = metadata["Individual Key"]
        model_id = metadata["Model Key"]
        biospecimen_annotations = list(zip(column_list, metadata.tolist()))
    else:
        print(f"Metadata not found for: {biospecimen_key}")
```

Bankso replied:
This is great, thank you for the suggestion!

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -95,25 +95,26 @@ def collect_biospecimen_annotations(syn, file_tuples: dict, specimen_info_tuple:
 
 
-def collect_record_annotations(syn, info_tuple: tuple[str, str, str], tuple_dict: dict):
+def collect_record_annotations(syn, file_tuples: dict, info_tuple: tuple[str, str, str], tuple_dict: dict):
```
Same suggestion as above for the variable name.

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -122,13 +123,14 @@ def collect_record_annotations(syn, info_tuple: tuple[str, str, str], tuple_dict
+                if tuple_dict[file] == id:
+                    annotations = list(zip(column_list, list(row)))
+                    apply_annotations_to_entity(syn, component, file, annotations)
+                    count += 1
```
Same suggestion as above for using pandas methods over the nested `for` loops.

vpchung reviewed (changes requested):
Made some suggestions! 😄 

But definitely take my suggestions with a grain of salt - my testing was done with very basic tables, so it's possible my code doesn't cover all the edge cases yours does.

Bankso committed:
Update table_to_annotations.py

Add changes suggested by @vpchung and lint with black

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -20,246 +20,287 @@ def get_args():
     component, table_id, column_list = specimen_info_tuple
-    data_table = get_table(syn, table_id, column_list)
+    data_table = get_table(syn, table_id, column_list).set_index("Biospecimen_id")
+    column_list.pop(0)
```
Maybe add a comment here to explain why you're popping the list?

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -20,246 +20,287 @@ def get_args():
+
     component, table_id, column_list = info_tuple
-    data_table = get_table(syn, table_id, column_list)
+    key_column = "_".join([component, "id"])
```
Just curious – is there a possibility this list of items to join together will expand in the future? If not, I suggest simplifying this to an f-string, which is more Pythonic and slightly faster.

```python
key_column = f"{component}_id"
```

Bankso replied:
Got it! There is a chance that this could be modified in the future, but I think it's good to use the f-string, for now.

vpchung commented on utils/table_to_annotations.py:
```diff
@@ -20,246 +20,287 @@ def get_args():
     filtered_annotations = [tup for tup in new_annotations if len(tup[1]) > 0]
-    for annot in filtered_annotations:
-        entity_annotations[annot[0].replace(" ", "")] = annot[1]
+    for key, annot in filtered_annotations:
```
I may be wrong, but I think an error will be encountered here, since `filtered_annotations` is a list, not a dictionary.

(Side note: if you want to unpack a dictionary into keys and values, you'll need to include `items()` as well)

Bankso replied:
> I think an error will be encountered here, since filtered_annotations is a list

It seemed to work for me! It is a list of tuples, in case that's relevant

vpchung replied:
Ah okay, great!!  I tested with a random synID and encountered errors, but so long as it works on your end, we're good :-)

vpchung reviewed (changes requested):
Sorry, a few more suggestions! :persevere:

Bankso committed:
Update table_to_annotations.py

- add in line comments to pop calls
- generate key column label using f-string instead of join()

vpchung reviewed (approved):
LGTM!
